lib/improver/cli/__init__.py

```diff
@@ -0,0 +1,405 @@
+        super().__init__(subject, owner, builder)
+
+
+# input handling
+
+
+class ObjectAsStr(str):
+    """Hide object under a string to pass it through Clize parser."""
+    __slots__ = ('original_object',)
+
+    def __new__(cls, obj, name=None):
+        if isinstance(obj, cls):  # pass object through if already wrapped
+            return obj
+        if name is None:
+            name = cls.obj_to_name(obj)
+        self = str.__new__(cls, name)
+        self.original_object = obj
+        return self
+
+    @staticmethod
+    def obj_to_name(obj, cls=None):
+        if cls is None:
+            cls = type(obj)
+        try:
+            obj_id = hash(obj)
+        except TypeError:
+            obj_id = id(obj)
+        return '<%s.%s@%i>' % (cls.__module__, cls.__name__, obj_id)
+
+
+def maybe_coerce_with(convert, obj, **kwargs):
+    """Apply converter if str, pass through otherwise."""
+    obj = getattr(obj, 'original_object', obj)
+    return convert(obj, **kwargs) if isinstance(obj, str) else obj
+
+
+@value_converter
+def input_cube(to_convert):
+    """
+
+    Args:
+        to_convert (string or obj):
+            calls maybe_coerce_with function with the input and load_cube.
+
+    Returns:
+        (obj):
+            The result of maybe_coerce_with.
+
+    """
+    from improver.utilities.load import load_cube
+    return maybe_coerce_with(load_cube, to_convert)
+
+
+@value_converter
+def input_json(to_convert):
+    """Calls maybe_coerce with a string or obj and returns.
+
+    Args:
+        to_convert (string or obj):
+            calls maybe_coerce_with function with the input and
+            load_json_or_none.
+
+    Returns:
+        (obj):
+            The result of maybe_coerce_with.
+
+    """
+    from improver.utilities.cli_utilities import load_json_or_none
+    return maybe_coerce_with(load_json_or_none, to_convert)
+
+
+# output handling
+
+
+@decorator
+def with_output(wrapped, *args, output=None, **kwargs):
+    """
+
+    Args:
+        wrapped (obj):
+            The function to be wrapped.
+        output (str, optional):
+            Output file name.
+    """
+    from improver.utilities.save import save_netcdf
+    result = wrapped(*args, **kwargs)
+    if output:
+        save_netcdf(result, output)
+        return
+    return result
+
+
+@decorator
+def with_intermediate_output(wrapped, *args, intermediate_output=None,
+                             **kwargs):
+    """
+    Args:
+        wrapped (obj):
+            The function to be wrapped.
+        intermediate_output (str, optional):
+            Output file name for intermediate result.
+    """
+
+    from improver.utilities.save import save_netcdf
+    result, intermediate_result = wrapped(*args, **kwargs)
+    if intermediate_output:
+        save_netcdf(intermediate_result, intermediate_output)
+    return result
+
+
+# cli object creation
+
+
+def _clizefy(obj, **kwargs):
+    # TODO: simplify after all CLIs are clizefied
```

Comment:
Some more explanation about how the current logic picks which interface to use
```suggestion
    # TODO: simplify after all CLIs are clizefied
    # The current logic allows for legacy argparse and new clize CLIs
    # to coexist. The legacy interface expects <cli_name>.main routine and
    # the new one expects <cli_name>.process routine that is type annotated.
    # If both interfaces are available, then one is picked based on the
    # IMPROVER_USE_CLIZE environment variable, default is the legacy one.
    # The environment setting has to be done before `import improver.cli`!
```